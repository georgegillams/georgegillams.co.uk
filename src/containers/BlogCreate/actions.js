import { createRoutine } from 'redux-saga-routines';

import { CREATE_BLOG } from './constants';

export const createBlog = createRoutine(CREATE_BLOG);
