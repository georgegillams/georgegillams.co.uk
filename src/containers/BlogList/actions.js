import { createRoutine } from 'redux-saga-routines';

import { LOAD_BLOGS, DELETE_BLOG } from './constants';

export const loadBlogs = createRoutine(LOAD_BLOGS);
export const deleteBlog = createRoutine(DELETE_BLOG);
