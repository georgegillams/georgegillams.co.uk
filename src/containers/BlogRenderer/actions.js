import { createRoutine } from 'redux-saga-routines';

import { LOAD_BLOG, UPDATE_BLOG } from './constants';

export const loadBlog = createRoutine(LOAD_BLOG);
export const updateBlog = createRoutine(UPDATE_BLOG);
