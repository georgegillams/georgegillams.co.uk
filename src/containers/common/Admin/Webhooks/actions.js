import { createRoutine } from 'redux-saga-routines';

import { LOAD_ENDPOINTS, CREATE_ENDPOINT, REMOVE_ENDPOINT, UPDATE_ENDPOINT } from './constants';

export const loadEndpoints = createRoutine(LOAD_ENDPOINTS);
export const createEndpoint = createRoutine(CREATE_ENDPOINT);
export const updateEndpoint = createRoutine(UPDATE_ENDPOINT);
export const removeEndpoint = createRoutine(REMOVE_ENDPOINT);
