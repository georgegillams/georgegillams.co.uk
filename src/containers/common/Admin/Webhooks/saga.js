import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';

import { selectState } from './selectors';
import {
  loadEndpoints,
  createEndpoint,
  updateEndpoint,
  removeEndpoint,
  loadNotifications,
  removeNotification,
} from './actions';

export function* doLoadEndpoint() {
  const requestURL = apiStructure.loadWebhookEndpoints.fullPath;

  try {
    yield put(loadEndpoints.request());

    const result = yield call(request, requestURL);

    if (result.error) {
      yield put(loadEndpoints.failure(result));
    } else {
      yield put(loadEndpoints.success(result.webhookEndpoints));
    }
  } catch (err) {
    yield put(loadEndpoints.failure(err));
  } finally {
    yield put(loadEndpoints.fulfill());
  }
}

export function* doCreateEndpoint() {
  const currentState = yield select(selectState());
  const { webhookEndpointToCreate } = currentState;
  const requestURL = apiStructure.createWebhookEndpoint.fullPath;

  try {
    yield put(createEndpoint.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(webhookEndpointToCreate),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(createEndpoint.failure(result));
    } else {
      yield put(createEndpoint.success(result));
      yield put(loadEndpoints.trigger());
    }
  } catch (err) {
    yield put(createEndpoint.failure(err));
  } finally {
    yield put(createEndpoint.fulfill());
  }
}

export function* doUpdateEndpoint() {
  const currentState = yield select(selectState());
  const { webhookEndpointToUpdate, onUpdateSuccessCb } = currentState;
  const requestURL = apiStructure.updateWebhookEndpoint.fullPath;

  try {
    yield put(updateEndpoint.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(webhookEndpointToUpdate),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(updateEndpoint.failure(result));
    } else {
      yield put(updateEndpoint.success(result));
      yield put(loadEndpoints.trigger());
      if (onUpdateSuccessCb) {
        onUpdateSuccessCb();
      }
    }
  } catch (err) {
    yield put(updateEndpoint.failure(err));
  } finally {
    yield put(updateEndpoint.fulfill());
  }
}

export function* doRemoveEndpoint() {
  const currentState = yield select(selectState());
  const { webhookEndpointToRemove } = currentState;
  const requestURL = apiStructure.deleteWebhookEndpoint.fullPath;

  try {
    yield put(removeEndpoint.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(webhookEndpointToRemove),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(removeEndpoint.failure(result));
    } else {
      yield put(removeEndpoint.success(result));
      yield put(loadEndpoints.trigger());
    }
  } catch (err) {
    yield put(removeEndpoint.failure(err));
  } finally {
    yield put(removeEndpoint.fulfill());
  }
}

export function* doLoadNotifications() {
  const currentState = yield select(selectState());
  const { webhookIdToLoadNotifications } = currentState;
  const requestURL = apiStructure.loadWebhookNotifications.fullPath.replace(':id', webhookIdToLoadNotifications);

  try {
    yield put(loadNotifications.request());

    const result = yield call(request, requestURL);

    if (result.error) {
      yield put(loadNotifications.failure(result));
    } else {
      yield put(loadNotifications.success(result));
    }
  } catch (err) {
    yield put(loadNotifications.failure(err));
  } finally {
    yield put(loadNotifications.fulfill());
  }
}

export function* doRemoveNotification() {
  const currentState = yield select(selectState());
  const { webhookIdToRemoveNotification } = currentState;
  const requestURL = apiStructure.deleteWebhookNotification.fullPath;

  try {
    yield put(removeNotification.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(webhookIdToRemoveNotification),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(removeNotification.failure(result));
    } else {
      yield put(removeNotification.success(result));
      yield put(loadNotifications.trigger(webhookIdToRemoveNotification.webhookId));
    }
  } catch (err) {
    yield put(removeNotification.failure(err));
  } finally {
    yield put(removeNotification.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(loadEndpoints.TRIGGER, doLoadEndpoint);
  yield takeLatest(createEndpoint.TRIGGER, doCreateEndpoint);
  yield takeLatest(updateEndpoint.TRIGGER, doUpdateEndpoint);
  yield takeLatest(removeEndpoint.TRIGGER, doRemoveEndpoint);
  yield takeLatest(loadNotifications.TRIGGER, doLoadNotifications);
  yield takeLatest(removeNotification.TRIGGER, doRemoveNotification);
}
