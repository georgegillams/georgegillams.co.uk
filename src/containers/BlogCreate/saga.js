import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';
import { selectState } from './selectors';

import { createBlog } from './actions';

export function* doCreateBlog() {
  const currentState = yield select(selectState());
  const { blogToCreate } = currentState;
  const requestURL = apiStructure.createBlog.fullPath;

  try {
    yield put(createBlog.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(blogToCreate),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(createBlog.failure(result));
    } else {
      yield put(createBlog.success(result));
      window.location = `/blog/${result.id}`;
    }
  } catch (err) {
    yield put(createBlog.failure(err));
  } finally {
    yield put(createBlog.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(createBlog.TRIGGER, doCreateBlog);
}
