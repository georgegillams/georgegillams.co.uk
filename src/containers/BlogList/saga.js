import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'utils/common/request';
import apiStructure from 'helpers/apiStructure';
import { selectState } from './selectors';

import { loadBlogs, deleteBlog } from './actions';

export function* doLoadBlogs() {
  const requestURL = apiStructure.loadBlogs.fullPath;

  try {
    yield put(loadBlogs.request());

    const result = yield call(request, requestURL, {
      method: 'GET',
    });

    if (result.error) {
      yield put(loadBlogs.failure(result));
    } else {
      yield put(loadBlogs.success(result));
    }
  } catch (err) {
    yield put(loadBlogs.failure(err));
  } finally {
    yield put(loadBlogs.fulfill());
  }
}

export function* doDeleteBlog() {
  const currentState = yield select(selectState());
  const { blogToDelete } = currentState;
  const requestURL = apiStructure.deleteBlog.fullPath;

  try {
    yield put(deleteBlog.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify({ id: blogToDelete }),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(deleteBlog.failure(result));
    } else {
      yield put(deleteBlog.success());
    }
    yield put(loadBlogs.trigger());
  } catch (err) {
    yield put(deleteBlog.failure(err));
  } finally {
    yield put(deleteBlog.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(loadBlogs.TRIGGER, doLoadBlogs);
  yield takeLatest(deleteBlog.TRIGGER, doDeleteBlog);
}
