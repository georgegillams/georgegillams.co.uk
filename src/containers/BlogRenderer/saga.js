import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'utils/common/request';
import apiStructure from 'helpers/apiStructure';

import { selectState } from './selectors';
import { loadBlog, updateBlog } from './actions';

export function* doLoadBlog() {
  const currentState = yield select(selectState());
  const { blogId } = currentState;
  const requestURL = apiStructure.loadBlog.fullPath.split(':id').join(blogId);

  try {
    yield put(loadBlog.request());

    const result = yield call(request, requestURL, {
      method: 'GET',
    });

    if (result.error) {
      yield put(loadBlog.failure(result));
    } else {
      yield put(loadBlog.success(result));
    }
  } catch (err) {
    yield put(loadBlog.failure(err));
  } finally {
    yield put(loadBlog.fulfill());
  }
}

export function* doUpdateBlog() {
  const currentState = yield select(selectState());
  const { blogToUpdate } = currentState;
  const requestURL = apiStructure.updateBlog.fullPath;

  try {
    yield put(updateBlog.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(blogToUpdate),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(updateBlog.failure(result));
    } else {
      yield put(updateBlog.success(result));
    }
    yield put(loadBlog.trigger(blogToUpdate.id));
  } catch (err) {
    yield put(updateBlog.failure(err));
  } finally {
    yield put(updateBlog.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(loadBlog.TRIGGER, doLoadBlog);
  yield takeLatest(updateBlog.TRIGGER, doUpdateBlog);
}
