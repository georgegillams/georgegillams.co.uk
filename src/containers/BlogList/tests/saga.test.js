import { put, takeLatest } from 'redux-saga/effects';

import { loadBlogs, deleteBlog } from '../actions';

import saga, { doLoadBlogs, doDeleteBlog } from '../saga';

describe('BlogList saga', () => {
  let mainSaga;

  beforeEach(() => {
    mainSaga = saga();
  });

  it('Should load blogs on loadBlogs TRIGGER', () => {
    const takeLatestDescriptor = mainSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(loadBlogs.TRIGGER, doLoadBlogs));
  });

  it('Should load blogs on deleteBlog TRIGGER', () => {
    mainSaga.next();
    const takeLatestDescriptor = mainSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(deleteBlog.TRIGGER, doDeleteBlog));
  });

  describe('loadBlogs actions', () => {
    let loadBlogsGenerator;

    const response = {
      blogs: 'list of blogs',
      status: 200,
    };

    beforeEach(() => {
      loadBlogsGenerator = doLoadBlogs();

      const selectDescriptor = loadBlogsGenerator.next().value;
      expect(selectDescriptor).toMatchSnapshot();
    });

    it('Should call loadBlogs.success on successful API call', () => {
      loadBlogsGenerator.next();
      const putSuccess = loadBlogsGenerator.next(response).value;
      loadBlogsGenerator.next();

      expect(putSuccess).toEqual(put(loadBlogs.success(response)));
    });

    it('Should call loadBlogs.failure if API returns error', () => {
      const response = {
        error: 'error_type',
        errorMessage: 'Error message',
        status: 500,
      };
      loadBlogsGenerator.next();
      const putSuccess = loadBlogsGenerator.next(response).value;
      loadBlogsGenerator.next();

      expect(putSuccess).toEqual(put(loadBlogs.failure(response)));
    });

    it('Should call loadBlogs.failure if an exception occurs', () => {
      const response = new Error('Some error');
      const putFailure = loadBlogsGenerator.throw(response).value;

      expect(putFailure).toEqual(put(loadBlogs.failure(response)));
    });
  });

  describe('deleteBlog actions', () => {
    let deleteBlogGenerator;

    const response = {
      status: 200,
    };

    beforeEach(() => {
      deleteBlogGenerator = doDeleteBlog();

      const selectDescriptor = deleteBlogGenerator.next().value;
      expect(selectDescriptor).toMatchSnapshot();
    });

    it('Should call deleteBlog.success on successful API call', () => {
      deleteBlogGenerator.next({ blogToDelete: 'some-blog-id' });
      deleteBlogGenerator.next();
      const putSuccess = deleteBlogGenerator.next(response).value;
      deleteBlogGenerator.next();

      expect(putSuccess).toEqual(put(deleteBlog.success()));
    });

    it('Should call deleteBlog.failure if API returns error', () => {
      const response = {
        error: 'error_type',
        errorMessage: 'Error message',
        status: 500,
      };
      deleteBlogGenerator.next({ blogToDelete: 'some-blog-id' });
      deleteBlogGenerator.next();
      const putSuccess = deleteBlogGenerator.next(response).value;
      deleteBlogGenerator.next();

      expect(putSuccess).toEqual(put(deleteBlog.failure(response)));
    });

    it('Should call deleteBlog.failure if an exception occurs', () => {
      const response = new Error('Some error');
      deleteBlogGenerator.next({ blogToDelete: 'some-blog-id' });
      const putFailure = deleteBlogGenerator.throw(response).value;

      expect(putFailure).toEqual(put(deleteBlog.failure(response)));
    });
  });
});
