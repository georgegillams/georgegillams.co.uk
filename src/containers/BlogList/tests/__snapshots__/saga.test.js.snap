// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogList saga deleteBlog actions Should call deleteBlog.failure if API returns error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`BlogList saga deleteBlog actions Should call deleteBlog.failure if an exception occurs 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`BlogList saga deleteBlog actions Should call deleteBlog.success on successful API call 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`BlogList saga loadBlogs actions Should call loadBlogs.failure if API returns error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "type": "app/blog-list/LOAD_BLOGS/REQUEST",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`BlogList saga loadBlogs actions Should call loadBlogs.failure if an exception occurs 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "type": "app/blog-list/LOAD_BLOGS/REQUEST",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;

exports[`BlogList saga loadBlogs actions Should call loadBlogs.success on successful API call 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "action": Object {
      "type": "app/blog-list/LOAD_BLOGS/REQUEST",
    },
    "channel": undefined,
  },
  "type": "PUT",
}
`;
